### PR TITLE
Added implementation for itemprop and itemtype along with test files

### PR DESCRIPTION
**Pull request by CJ8664**
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

- [x] There are tests for these changes OR

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19038)

### CODE REVIEW

**Comment by CJ8664** [comment]
@KiChjang , @fitzgen 
The build error says `error: no rules expected the token `"itemtype"`
   --> C:\projects\servo\components\script\dom\htmlelement.rs:316:26
    |
316 |             &local_name!("itemtype"),
    |                          ^^^^^^^^^^
error: no rules expected the token `"itemprop"`
   --> C:\projects\servo\components\script\dom\htmlelement.rs:346:26
    |
346 |             &local_name!("itemprop"),
    |                          ^^^^^^^^^^
error: no rules expected the token `"itemprop"`
   --> C:\projects\servo\components\script\dom\htmlelement.rs:696:26
    |
696 |             &local_name!("itemprop") => AttrValue::from_serialized_tokenlist(value.into()),
    |                          ^^^^^^^^^^
error: no rules expected the token `"itemtype"`
   --> C:\projects\servo\components\script\dom\htmlelement.rs:697:26
    |
697 |             &local_name!("itemtype") => AttrValue::from_serialized_tokenlist(value.into()),
    |                          ^^^^^^^^^^
error: aborting due to 4 previous errors
error: Could not compile `script`.
`

Whereas we had merged a pull request to add these into html5ever [https://github.com/servo/html5ever/pull/320/files](https://github.com/servo/html5ever/pull/320/files)

**Comment by KiChjang** [comment]
Have you ran `./mach update-cargo -p markup5ever`?

Also, merge commits are not accepted. Please rebase against the master branch.

**Comment by CJ8664** [comment]
Closing pull request to make minor tidy modifications

File: tests/wpt/mozilla/meta/MANIFEST.json
Comment (KiChjang) [inline tests/wpt/mozilla/meta/MANIFEST.json]: This... doesn't seem it should be here.

File: tests/wpt/mozilla/meta/MANIFEST.json
Comment (KiChjang) [inline tests/wpt/mozilla/meta/MANIFEST.json]: Same here.

**Comment by CJ8664** [comment]
@KiChjang we are having some difficulties in rebasing our branch, is there a possibility/option to squash the commits while merging?

**Comment by KiChjang** [comment]
I can help you rebase against master, if you have the "Allow maintainers to edit branch" checkbox checked at the right hand side of GitHub's interface.

Please let me know whether this is preferable to you, and if you agree to it, please do not push any code to your branch until my rebasing is complete.

File: components/script/dom/webidls/HTMLElement.webidl
Comment (KiChjang) [inline components/script/dom/webidls/HTMLElement.webidl]: Why are these commented out?

Comment (niravjain) [reply components/script/dom/webidls/HTMLElement.webidl]: I think these lines were already there. We don't need these comments, I can delete them

File: .gitignore
Comment (jdm) [inline .gitignore]: I assume this is a script you wrote for your own testing purposes? Let's not add it to .gitignore; you can just leave it as an untracked file.

File: components/script/dom/webidls/HTMLElement.webidl
Comment (jdm) [inline components/script/dom/webidls/HTMLElement.webidl]: Since these are only for use in test code, let's use the dom.microdata.testing.enabled name instead.

File: resources/prefs.json
Comment (jdm) [inline resources/prefs.json]: dom.microdata.testing.enabled

File: tests/wpt/mozilla/meta/mozilla/microdata/__dir__.ini
Comment (jdm) [inline tests/wpt/mozilla/meta/mozilla/microdata/__dir__.ini]: dom.microdata.testing.enabled

File: tests/wpt/mozilla/tests/mozilla/microdata/dup_prop_type_test.html
Comment (jdm) [inline tests/wpt/mozilla/tests/mozilla/microdata/dup_prop_type_test.html]: Let's reduce some of the duplication in this test (and others) like this:
```js
var ids = ["body", "h1", "h2", "p", "input", "form", "li", "ul", "meta", "table", "div", "anchor"];
for (let id of ids) {
    var elem = document.getElementById(id);
    assert_array_equals(elem.propertyNames().sort(), ["hello", "world"], "duplicate itemprop tokens should be filtered");
    assert_array_equals(elem.itemtypes().sort(), ["hello", "world"], "duplicate itemtype tokens should be filtered");
}
```

File: components/script/dom/htmlelement.rs
Comment (jdm) [inline components/script/dom/htmlelement.rs]: This file contains a lot of modifications caused by your editor automatically running rustfmt on the file, which makes it harder to review. Could you disable this feature, revert the changes to the file, and make your modifications to it again?

File: components/script/dom/htmlelement.rs
Comment (jdm) [inline components/script/dom/htmlelement.rs]: Rather than calling parse_plain_attribute here, we should use [Element::get_tokenlist_attribute](http://doc.servo.org/script/dom/element/struct.Element.html#method.get_tokenlist_attribute) instead, which will return the previously-parsed value.

File: components/script/dom/htmlelement.rs
Comment (jdm) [inline components/script/dom/htmlelement.rs]: Instead of manually draining and pushing, we can replace this with:
```rust
Some(item_attrs.into_iter().collect())
```

File: components/script/dom/htmlelement.rs
Comment (jdm) [inline components/script/dom/htmlelement.rs]: Same comments for this method as the previous one.

**Review by jdm — CHANGES_REQUESTED** [review]
Good work!

**Comment by KiChjang** [comment]
@CJ8664 Actually, the better path forward here is to address @jdm's comments first - once we are OK with your changes, we'll rebase this on our side. Sorry for the confusion!

**Comment by CJ8664** [comment]
Sure @KiChjang , we will incorporate the suggestions given by @jdm and then rebase-> merge. @jdm Thanks for the suggestions, as we are new to Rust, we will keep on improving with further develo[ment and reviews.

**Comment by vjhebbar** [comment]
Thank you for the review comments. We will begin incorporating changes and provide an update.

**Comment by CJ8664** [comment]
Closing the pull request, to make changes. Will reopen once the changes are pushed

**Comment by jdm** [comment]
It is perfectly acceptable to leave the PR open while you make changes.

**Comment by CJ8664** [comment]
Changes other than htmlelement.rs made and repo synced to resolve conflicts. Opening pull request to check integration pass

File: components/script/dom/htmlelement.rs
Comment (jdm) [reply components/script/dom/htmlelement.rs]: This does not appear to have been addressed.

**Comment by niravjain** [comment]
@jdm How do we manually trigger checks?

**Comment by jdm** [comment]
What checks are you referring to?

**Comment by niravjain** [comment]
The continuous-integeration fails with the error "AppVeyor was unable to build non-mergeable pull request", however we're not seeing any conflicts when we do a "git pull"

**Comment by CJ8664** [comment]
HI @jdm, usually we see what are the issues in the Integration build, however, now we are not able to see what is the issue nor manually trigger an appveyor build

**Comment by jdm** [comment]
Ignore the appveyor problem; github and appveyor have been having some issues some PRs today.

File: components/script/dom/htmlelement.rs
Comment (CJ8664) [reply components/script/dom/htmlelement.rs]: About this change,  do we then remove the parse_plain_attribute method that we implemented ? because its purpose was to be used in this function

Comment (jdm) [reply components/script/dom/htmlelement.rs]: No, parse_plain_attribute is a method that gets called from other parts of Servo already. Your implementation is required, like adding a virtual method override in C++.

Comment (CJ8664) [reply components/script/dom/htmlelement.rs]: Got it, we have now used Element::get_tokenlist_attribute and will push the code in few minutes, running test-tidy now. When can we expect the appveyor to work again? ;(

Comment (KiChjang) [reply components/script/dom/htmlelement.rs]: AppVeyor passing is not required for landing. Also, when we finally approve your changes and are ready to land your changes, we have another build & test system that will build and test your changes against all web platform tests before merging your PR to the master branch.

**Comment by jdm** [comment]
Note: when you modify a test file, you need to run `./mach update-manifest` and commit the results before pushing your changes, or TravisCI will complain.

**Comment by CJ8664** [comment]
@jdm is appveyor working fine now? If so how can we manually trigger a build

**Comment by KiChjang** [comment]
@CJ8664 I think my message got lost in the comments, but AppVeyor is not a strict requirement to land your PR. However, if you are concerned on whether your code builds and tests correctly in Windows, we can use @bors-servo try to perform a test & build run across all 3 major platforms.
Hopefully everything passes fine!

**Comment by CJ8664** [comment]
@KiChjang, this is great news!! in this case can you help us squash commits and merge the PR? Actually, this project is for our coursework at college and the grading kinda depends on the status of pull request :) Once this phase is complete we have to start with the next phase. About the build it works fine on our machines and tests also runs correctly

**Comment by KiChjang** [comment]
@jdm Is this good enough to be merged, or is there any outstanding issue that you'd like the team to address?

File: components/script/dom/htmlelement.rs
Comment (jdm) [inline components/script/dom/htmlelement.rs]: This file still contains many formatting changes that are unrelated to the new APIs being implemented in it.

File: components/script/dom/htmlelement.rs
Comment (jdm) [inline components/script/dom/htmlelement.rs]: I previously recommended this:
```rust
fn Itemtypes(&self) -> Option<Vec<DOMString>> {
    if atoms.is_empty() {
        return None;
    }

    let mut item_attr_values = HashSet::new();
    for attr_value in &atoms {
        item_attr_values.insert(DOMString::from(String::from(attr_value.trim())));
    }

    Some(item_attr_values.into_iter().collect())
}
```

File: components/script/dom/htmlelement.rs
Comment (jdm) [inline components/script/dom/htmlelement.rs]: Same as previous comment.

**Review by jdm — CHANGES_REQUESTED** [review]
Just three unaddressed comments from the previous review.

**Comment by CJ8664** [comment]
Hey, @bors-servo can you try to perform a test & build run across all 3 major platforms.

**Comment by CJ8664** [comment]
Trolled by a bot :P

**Comment by CJ8664** [comment]
@KiChjang @jdm could you review the changes now? I have picked the htmlelement.rs file from servo base and included only our changes, thus this time the editor made changes should not be there. Also dried out the methods

**Comment by jdm** [comment]
Looks good to me; thanks!

**Comment by KiChjang** [comment]
I've taken the liberty to rebase all the changes on this branch, and made a new branch on my own fork: https://github.com/KiChjang/servo/tree/student. @jdm Should I squash that branch, and then force-push onto this one?

**Comment by jdm** [comment]
That would be great!

**Comment by CJ8664** [comment]
@KiChjang for you to be able to this, shall I add you as a collaborator in the forked repo?

**Comment by KiChjang** [comment]
@CJ8664 Not necessary; on the GitHub UI, if you have the "Allow maintainer to edit branch" checkbox checked, then people with write access to the servo repository would be able to push to your PR branch. I've just done that, so you should see all your team's work squashed into 1 commit on this branch.

Thank you for your work!

@bors-servo r=jdm

**Comment by jdm** [comment]
Sigh. There's an extra newline at the end of Cargo.lock that needs to be removed. :(

**Comment by CJ8664** [comment]
Hey @KiChjang should we remove the extraline from Cargo.lock or pushing our changes onto this PR would mess it up?

**Comment by jdm** [comment]
```
$ git fetch origin master
$ git checkout master
$ git reset --hard 7b3794b
... make the change ...
$ git commit -a --amend --no-edit
$ git push -f origin master
```
That should take care of it.

**Comment by CJ8664** [comment]
Lets see if it builds this time

**Review by jdm — APPROVED** [review]

**Pull request merged**
